{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modules and Calling External Programs\n",
    "## Topics\n",
    "- Modules - indepedent collections of related functions\n",
    "- Using Modules - math & collections\n",
    "- Making your own module\n",
    "- Interacting with other software using __subprocess__\n",
    "- Making your own BLAST module\n",
    "\n",
    "\n",
    "## Introduction\n",
    "\n",
    "We're covering a lot of ground today!\n",
    "\n",
    "First up we'll be discussing modules and how they are used to organize code.\n",
    "\n",
    "Then we're getting into ways to call external programs from within Python. We'll also cover saving the output of these programs, either to new files or within python's memory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules\n",
    "\n",
    "In all of the examples so far, we defined our functions right above the code that we hoped to execute. If you have many functions, you can see how this would get messy in a hurry. \n",
    "Furthermore, part of the benefit of functions is that you can call them multiple times within a program to execute the same operations without writing them all out again. \n",
    "\n",
    "But wouldn't it be nice to share functions across programs, too? \n",
    "\n",
    "For example, working with genomic data means lots of time getting sequence out of FASTA files, and shuttling that sequence from program to program. Many of the programs we work with overlap to a significant degree, as they need to parse FASTA files, calculate evolutionary rates, and interface with our lab servers, for example -- all of which means that many of them share functions. And if the same function exists in two or more different programs, we hit the same problems that we hit before: complex debugging, decreased readability, and, of course, too much typing.\n",
    "\n",
    "Modules solve these problems. In short, they're collections of functions and variables (and often objects, which we'll get to towards the end of the course) that are kept together in a single file that can be read and __import__ed by any number of programs.\n",
    "\n",
    "### The Basics: Using the math module\n",
    "\n",
    "To illustrate the basics, we'll go through the use of the __math__ module, a module which we use almost all the time. To use a function or variable in the __math__ module, use the syntax \n",
    "__math.__*NameOfThing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##to use a module in your code, first import it\n",
    "import math\n",
    "\n",
    "x = 5\n",
    "\n",
    "##Modules usually contain functions\n",
    "log10 = math.log10(x)\n",
    "cos = math.cos(x)\n",
    "\n",
    "##sometimes modules contain variables\n",
    "pi = math.pi\n",
    "e = math.e\n",
    "\n",
    "print log10, cos, pi, e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "math.cos?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The collections module\n",
    "\n",
    "Another useful module is the __collections__ module. It has a bunch of new data types that are, as you might guess from the name, collections of other things. We will cover two of the most commonly used objects: __Counter__ and __defaultdict__. Let's start with __Counter__, which counts things."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import collections\n",
    " \n",
    "my_genera = ['Helicobacter', 'Escherichia', 'Lactobacillus',\n",
    "             'Lactobacillus', 'Oryza', 'Wolbachia', 'Oryza',\n",
    "             'Rattus', 'Lactobacillus', 'Drosophila']\n",
    " \n",
    "c = collections.Counter(my_genera)\n",
    "print c\n",
    "##Note that placing the list into Counter() immediately gets\n",
    "##you the count."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The collections module gives us a new data type, __Counter__, that counts things. It is essentially a dictionary where the key is some element we are recording and the value is the count of how often it appears. Remember that list of amino acids we got the count for in the exercises in Section 2.1? There, we created a dictionary where every key was initialized with a value of zero, and then proceeded to add one for each observance. Here, we can just use a __Counter__ to get the count of each unique element in the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##This is how we did a count in a dictionary. Many more lines of code!\n",
    "counts = {}\n",
    " \n",
    "for genus in my_genera:\n",
    "    if genus not in counts:\n",
    "        counts[genus] = 0\n",
    "    counts[genus] += 1\n",
    "\n",
    "print \"The dictionary\", counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using a __Counter__ is faster to write and saves us writing this bit of code every time we want to something. Another big advantage of the __Counter__ type is that it makes it really easy to sort by frequency:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "my_seq = ['MET', 'GLU', 'VAL', 'LYS', 'ARG', 'GLU', 'HIS', 'TRP', 'ALA',\n",
    "          'THR', 'ARG', 'LEU', 'GLY', 'LEU', 'ILE', 'LEU', 'ALA', 'MET',\n",
    "          'ALA', 'GLY', 'ASN', 'ALA', 'VAL', 'GLY', 'LEU', 'GLY', 'ASN',\n",
    "          'PHE', 'LEU', 'ARG', 'PHE', 'PRO', 'VAL', 'GLN', 'ALA', 'ALA',\n",
    "          'GLU', 'ASN', 'GLY', 'GLY', 'GLY', 'ALA', 'PHE', 'MET', 'ILE',\n",
    "          'PRO', 'TYR', 'ILE', 'ILE', 'ALA', 'PHE', 'LEU', 'LEU', 'VAL',\n",
    "          'GLY', 'ILE', 'PRO', 'LEU', 'MET', 'TRP', 'ILE', 'GLU', 'TRP',\n",
    "          'ALA', 'MET', 'GLY', 'ARG', 'TYR', 'GLY', 'GLY', 'ALA', 'GLN',\n",
    "          'GLY', 'HIS', 'GLY', 'THR', 'THR', 'PRO', 'ILE', 'VAL', 'PHE',\n",
    "          'LEU', 'ILE', 'THR', 'MET', 'PHE', 'ILE', 'ASN', 'VAL', 'SER',\n",
    "          'ILE', 'LEU', 'ILE', 'ARG', 'GLY', 'ILE', 'SER', 'LYS', 'GLY',\n",
    "          'ILE', 'GLU', 'ARG', 'PHE', 'ALA', 'LYS', 'ILE', 'ALA', 'MET',\n",
    "          'PRO', 'THR', 'LEU', 'PHE', 'ILE', 'LEU', 'ALA', 'VAL', 'PHE',\n",
    "          'LEU', 'VAL', 'ILE', 'ARG', 'VAL', 'PHE', 'LEU', 'LEU', 'GLU',\n",
    "          'THR', 'PRO', 'ASN', 'GLY', 'THR', 'ALA', 'ALA', 'ASP']\n",
    "\n",
    "c = collections.Counter(my_seq)\n",
    " \n",
    "print c\n",
    "print\n",
    "print\n",
    "print c.most_common()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Counter.*__most_common()__ returns a list of tuples, sorted in order by highest count to lowest count.\n",
    "\n",
    "The other __collections__ type we will cover is __defaultdict__, which is also like a dictionary, but has a default type for a key that we haven't seen before (with a normal dictionary, if you try to read something where the key isn't in the dict, then you get an error). Let's think about how we'd make a dictionary where each key is a genus, and the value is a list of species in that genus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import collections\n",
    " \n",
    "my_species = [('Helicobacter','pylori'), ('Escherichia','coli'),\n",
    "              ('Lactobacillus', 'helveticus'),\n",
    "              ('Lactobacillus', 'acidophilus'),\n",
    "              ('Oryza', 'sativa'), ('Wolbachia', 'pipientis'),\n",
    "              ('Oryza', 'glabberima'), ('Rattus', 'norvegicus'),\n",
    "              ('Lactobacillus','casei'), ('Drosophila','melanogaster')]\n",
    " \n",
    "# Below, we put the list into a normal dictionary, \n",
    "# with genera as keys and species as values\n",
    "d1 = {}\n",
    "for genus, species in my_species:\n",
    "    if genus not in d1:\n",
    "        d1[genus] = []\n",
    "    d1[genus].append(species)\n",
    "\n",
    "print \"normal dictionary -- \", d1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With a __defaultdict__, we can once again save the line in the for loop where we check for a non-existent key:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "d2 = collections.defaultdict(list)\n",
    " \n",
    "for genus, species in my_species:\n",
    "    d2[genus].append(species)\n",
    "\n",
    "print \"default dict -- \", d2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Moreover, if we check for the species in a genus that has no species, we no longer receive an error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "print d2['Saccharomyces']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One thing to look at is the line where we actually declare the defaultdict: here we've given it another type, and if we use a key that's not in the dictionary already, it will initialize it to be an empty variable of that type. Most often, this will be a list, but you could imagine uses for other types, like a string, an integer (here \"empty\" actually would mean 0), or even another dict. It's possible to even have a defaultdict of defaultdicts!\n",
    "\n",
    "## Making Your Own Modules\n",
    "\n",
    "Now that you know the basics of __import__ing and using a module, you will learn how to write our own modules, which is almost just as easy!\n",
    "\n",
    "Any file of python code with a *.py* extension can be __import__ed as a module from your script. When you invoke an __import__ operation from a program, all the statements in the __import__ed module are executed immediately. The program also gains access to names assigned in the module (names can be functions, variables, classes, etc.), which can be invoked in the program using the syntax *module.name*. Find the following script in the file *greeting_module.py*:\n",
    "\n",
    "```python\n",
    "print 'The top of the greeting_module has been read.'\n",
    " \n",
    "def hello(name):\n",
    " greeting = \"Hello {}!\".format(name)\n",
    " return greeting\n",
    " \n",
    "def ahoy(name):\n",
    " greeting = \"Ahoy-hoy {}!\".format(name)\n",
    " return greeting\n",
    " \n",
    "x = 5\n",
    " \n",
    "print 'The bottom of the greeting_module has been read.'\n",
    "```\n",
    "\n",
    "The following script will call the module **greeting_module** and use the functions and variables located within the module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Make sure you reset your kernel before running this cell,\n",
    "# or you won't see the output from greeting_module.\n",
    "import greeting_module\n",
    "\n",
    "greeting = greeting_module.hello('Christopher')\n",
    "print greeting\n",
    "print\n",
    "\n",
    "x = 1\n",
    "print 'x within greeting_module:', greeting_module.x\n",
    "print 'x within the __main__ module:', x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that it runs through all of greeting module first, so anything that is printed out in *greeting_module.py* is also printed out before anything in test.py is run.\n",
    "\n",
    "And that's it! See-- no more messy function declarations at the beginning of your script. Now if you need any other program to say \"hi\" to you, all you need to do is __import__ the greeting module.\n",
    "\n",
    "---\n",
    "## Informative Interlude: The Kernel\n",
    "\n",
    "A kernel is a program that translates instructions from a user (or storage media) into machine operations, and produces output back to the user (or other program). The IPython Kernel that we are using for this notebook reads Python code that is given to it and translates it into instructions for the computer, then prints any output text. The vanilla Python Kernel does the same thing, but doesn't have fancy features like tab-completion and multi-line input.\n",
    "\n",
    "Most of the time, when you run a Python program, the Python Kernel reads a script (the _.py_ file) and exits when reaching an Exception, an exit instruction, or the end of the file. When you run python on the terminal by typing \"python\" or \"ipython\", instead of reading a script, the kernel reads from your direct input (called stdin). After reading a line you type in, this \"interactive kernel\" (actually an interpreter) pauses and waits for another input, rather than simply exiting, even if the last input caused an exception.\n",
    "\n",
    "In these Jupyter notebooks, we are using the IPython Kernel in a third way. It is as if we have started a Python interpreter by typing \"ipython\" in a terminal, and it is just sitting in the background waiting. When we run a cell of Python code, it is as if we copy that cell and paste it into our idle interpreter.\n",
    "\n",
    "This is why you need to restart your kernel before running cells with __import__ statements to make sure those imports worked correctly. Python is lazy and if it sees you are trying to __import__ a module that has already been __import__ed it will skip it. This may have also been the cause of some of your confounding errors throughout this class. If you ran cells out of order, _x_ may not have been what you thought it was. Or maybe you got stuck somewhere, unable to run any code, because some other cell was caught in an infinite loop, or waiting for some __raw_input()__.\n",
    "\n",
    "Here's where you can [read up on how the IPython Kernel actually works](https://ipython.org/ipython-doc/3/development/how_ipython_works.html \"How IPython works\") in Jupyter Notebooks.\n",
    "\n",
    "---\n",
    "\n",
    "## Using Modules Beyond 'import'\n",
    "\n",
    "Here, we will cover how to modify your __import__ statement to make using modules less clunky in your code.\n",
    "\n",
    "First, what if you only want one function from a given module? Let's say, as an Alexander Graham Bell loyalist, you really only dealt in 'ahoys' rather than 'hellos.' We need to use a modified syntax for retrieving _only_ the __ahoy__ function from the module, without cluttering things up by loading the newfangled __hello__ function preferred by T.A. Edison's entourage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Again, make sure to restart your kernel before running this.\n",
    "from greeting_module import ahoy\n",
    "\n",
    "greeting = ahoy('everybody')\n",
    "# if you grab a function from a module with a 'from' statement,\n",
    "# you don't need to use the <module>.<function> syntax\n",
    "print greeting\n",
    "print\n",
    "print hello('Christopher')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The __from__ _module_ __import__ _object_ syntax has two differences from just plain __import__ _module_. First we see that we can now write __ahoy('everybody')__ directly, instead of having to write __greeting_module.ahoy('everybody')__. Second, we don't have access to __hello__ or any of the other objects in the greeting module. If we wanted to access both functions this way, we could __import__ them both in one statement like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Restart your kernel\n",
    "from greeting_module import ahoy, hello\n",
    "print hello(\"You\")\n",
    "print ahoy(\"Matey\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or, what if there were a lot of functions from the __greeting_module__ we wanted to use, but didn't want to write out the full name? Rather than writing out all of the function names to import individually (there could be a lot of them), we can use the asterisk wildcard (_*_) to refer to them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Restart your kernel\n",
    "x = 1\n",
    "from greeting_module import *\n",
    "\n",
    "print x\n",
    "\n",
    "print ahoy('everybody')\n",
    "print hello('everybody')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While this may be useful if we are familiar with the contents of the module, including all of the names inside, there are a few reasons to be careful about using the __from__ _module_ __import *__ syntax. \n",
    "\n",
    "First, if the module contains a lot of variables that we don't need to use, we will needlessly allocate memory to storing the information.\n",
    "\n",
    "Second, and perhaps more importantly, if the module being imported contains variables with the same names as those inside your program, you will lose access to the original values of those variables. This is what happens to _x_, above.\n",
    "\n",
    "### Changing Names When Importing\n",
    "\n",
    "It is possible to avoid name conflicts by changing the name of objects you import using the __as__ keyword."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The top of the greeting_module has been read.\n",
      "The bottom of the greeting_module has been read.\n",
      "Hello World!\n",
      "What's up, Doc?\n"
     ]
    }
   ],
   "source": [
    "# Restart your kernel\n",
    "from greeting_module import hello as hi\n",
    "\n",
    "def hello(name):\n",
    "    greeting = \"What's up, Doc?\"\n",
    "    return greeting\n",
    "\n",
    "print hi('World')\n",
    "print hello('World')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This can also be useful to shorten commonly used module names when you want many things in the module, but don't want to run into conflicts by using __from__ _module_ __import *__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "import greeting_module as gm\n",
    "\n",
    "print gm.x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Where to Store Your Modules: using PYTHONPATH\n",
    "\n",
    "Over time, you'll end up accumulating lots of these modules, and they'll tend to fall together in meaningful collections.\n",
    "\n",
    "For example, \n",
    "- You might have a module for all your functions related to reading and parsing files, called *files_tools.py*. \n",
    "- You might have another for common sequence-related tasks, called *sequence_tools.py*. \n",
    "\n",
    "Python keeps its modules installed in a system directory.  If you are running Python on a remote server, you may not even have permissions to add files to this directory. Therefore, it's useful and simpler to just create your own Python modules directory and then let your operating system environment know about it. \n",
    "\n",
    "### Adding a folder to your Python Path: Linux and OSX\n",
    "Python doesn't look through every possible file and folder on your computer to find saved modules. Rather, it looks at a select set of folders, which are stored in your system's **PYTHONPATH** variable. You can add a new folder for Python to look in by adding a few lines to your *.bashrc* file in your home directory with the following terminal commands:\n",
    "```bash\n",
    "sudo nano ~/.bashrc\n",
    "```\n",
    "This will open your *.bashrc* file, which contains a set of bash commands run whenever you start a shell (terminal) to setup your environment. Add the following lines to the end of the file, substituting the path to your module. \n",
    "\n",
    "```bash\n",
    "PYTHONPATH=$PYTHONPATH:/path/to/your/module\n",
    "\n",
    "export PYTHONPATH\n",
    "```\n",
    "\n",
    "The PYTHONPATH variable is the list of all folders that Python knows contain modules. The above sets the PYTHONPATH variable (variables are indicated by a \"$\" in bash) to include the folder your module is in.\n",
    "\n",
    "Hit CONTROL + X to quit, then type 'y' and hit enter to save.\n",
    "\n",
    "Next, reload your *.bashrc*\n",
    "\n",
    "```bash\n",
    "source ~/.bashrc\n",
    "```\n",
    "\n",
    "And check if it worked using **echo**\n",
    "\n",
    "```bash\n",
    "echo $PYTHONPATH\n",
    "```\n",
    "\n",
    "And with that, any file that ends up in this directory will be treated as a module by Python. And though this is a good final resting place for your polished modules, you can also prototype them by simply saving them in your current working directory, and moving them over when you're happy with them.\n",
    "\n",
    "### Adding a folder to your Python Path: Windows\n",
    "\n",
    "On Windows, you can do this using the command prompt, but it's safer to use the built in graphical interface.\n",
    "\n",
    "To get there:\n",
    "- Open up \"My Computer\" or \"This PC\"\n",
    "- Right-click in the background and select \"Properties\"\n",
    "- On the \"Advanced\" tab, click the \"Environment Variables\" button\n",
    "- Add or edit your PYTHONPATH variable, seperating directories with a semicolon \";\" as seen below\n",
    "\n",
    "![Environment Variables](WindowsEnvVars.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# System Calls With the subprocess Module\n",
    "\n",
    "In this section we will cover how to use the **subprocess** module to interact with other software, making system calls, collecting data, and automating your entire analysis pipeline.\n",
    "\n",
    "In bioinformatics we will often want to use other peoples software, whether that's Python packages like Biopython or packages written in other languages, like the highly efficient C. Rather than reinvent the wheel, we use many specialized packages to quickly and easily perform tasks such as alignment, SNP calling, or phylogenetics. Even simple Unix tools like **wc** can be useful to count the number of genes in a GFF file or length of a FASTA sequence.\n",
    "\n",
    "Most bioinformatics tools, like Unix tools, are run through the command line. Given that you might want to repeat an analysis using some of these tools on, for example, hundreds of genes, manually entering the commands for your entire pipeline is not only boring and error prone but a waste of your time.\n",
    "\n",
    "Here we will teach you to automate your pipeline by using Python to run these command line tools for you.\n",
    "\n",
    "---\n",
    "## Advantages to Automation\n",
    "\n",
    "**Reproducability**\n",
    "\n",
    "Perhaps the greatest advantage to a scientist in automating analysis is that the analysis can be reproduced exactly. Your exact methods are laid out in your Python script, where you and others can scrutinize, repeat, and modify them.\n",
    "\n",
    "**Less Tedium**\n",
    "\n",
    "From printing out the name of every gene expressed over a certain level, to BLASTING those genes against the NCBI database, to sorting and counting the resulting hits, scripting saves you a huge amount of tedious labor. Nobody wants to type, or even copy and paste, hundreds of BLAST queries.\n",
    "\n",
    "**Consistency**\n",
    "\n",
    "As well as being incredibly mind-numbing, manually running bioinformatics tools is dangerous. What if you accidentally type 'Gasterosteus_aculeatus_CA_SNPs.vcf' instead of 'Gasterosteus_aculeatus_AC_SNPs.vcf', accidentally substituting your California population for your Atlantic Coast population? Or 'clean_reads.py expensive_dataset.fq > expensive_dataset.fq' instead of 'clean_reads.py expensive_dataset.fq > expensive_dataset.clean.fq'? There are *thousands* of ways you can accidentally screw up your analysis to either ruin your day or produce erroneous results.\n",
    "\n",
    "Automation reduces the risk of stupid typos and other accidents. You won't forget to include mydata.part.14.bam in the analysis when you run **results = [analyse(data) for data in mydata]**.\n",
    "\n",
    "**Parallelization**\n",
    "\n",
    "Modern computers, even budget laptops, now have multiple processors, which means you can run several or even hundreds of analyses at once (if you have access to a supercomputing cluster)! Python provides a number of tools to help you manage these processes and make the most out of parallel computing.\n",
    "\n",
    "---\n",
    "\n",
    "## subprocess.check_output()\n",
    "The easiest way to make system calls is with the __check_output()__ function in the __subprocess__ module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.2_Solutions.ipynb\n",
      "greeting_module.py\n",
      "Lecture 4.2 Modules and Calling External Programs.ipynb\n",
      "Popen.ipynb\n",
      "test.fasta\n",
      "test_output.py\n",
      "WindowsEnvVars.png\n",
      "yeast.aa\n",
      "yeast.nt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    " \n",
    "output = sp.check_output('ls', shell=True)\n",
    "\n",
    "print output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we used __subprocess.check_output()__ to run the command **ls**, and captured the output in the *output* variable.\n",
    "\n",
    "You'll notice that **check_output()** takes one mandatory argument: the command you want to run as you would type it into your terminal shell. The __shell__ keyword argument we will leave as __True__ for this tutorial. Without it, the process is created directly by the operating system, and any symbols or commands that the shell would recognize (e.g. spaces, \">\", and \"|\") result in an error. Because spaces are a symbol recognized by the shell, when calling __call()__ without __shell=True__, the first argument should be a list of command line arguments that would be separated by spaces spaces on the terminal. So **ls -a -1 /home/james** would become **['ls', '-a', '-1', '/home/james']**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Redirecting Output to Files\n",
    "Since we are using __shell=True__, you can redirect the output of a command to file exactly as you would on the terminal.\n",
    "\n",
    "To demonstrate this, we will be asking Python to do something very meta: run another Python script! Here's the script we will be running, which should be saved in your 5.2 directory as 'test_output.py'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting test_output.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile test_output.py\n",
    "#!/usr/bin/env python\n",
    "print 'this is a test'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember that we can redirect the output of a command to a file with __>__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "command = 'python test_output.py > out.txt'\n",
    "output = sp.check_output(command, shell=True)\n",
    "print output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This time there is no output to print, since we redirected it. Check 'out.txt' to make sure the output went where you expected it.\n",
    "\n",
    "### Chaining Commands With Pipes\n",
    "We often want to do something else with the output of a program, either parsing it and reformating it, performing a second step in the analysis, or turning it into a figure. To do this, we will *pipe* the output of one command directly into another command.\n",
    "\n",
    "A *pipe* is used to send the output of one program into the input of another. We learned in the first lecture that this is done on the Unix command line with the **|** character. For example, this script should count the number of files that contain the word \"yeast\" in the current directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "command = \"ls -1 | grep 'yeast' | wc -l\"\n",
    "print sp.check_output(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If these Unix commands are new to you, try breaking this chained command down to figure out what each step is doing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLAST\n",
    "Good ol'e Basic Local Alignment Search Tool, now that's bioinformatics even Grandma can get down with. In its most basic form, BLAST takes a short sequence of either amino acids or nucleotides, searches a database of reference sequences, and returns the most likely matches for the query sequence. It's brilliant, and the only trouble is that we're often interested in BLASTing hella query sequences, and much less interested in the mundanity of parsing the output that comes back to us. Fortunately, we can easily script the execution of BLAST such that we can run queries many times with different sequences and settings, and also parse the output back into dictionaries to manipulate and use for downstream analysis.\n",
    "\n",
    "The first part of this section will show us how to install a new program in our UNIX environment, and the second part will let us practice using **subprcess** to script the execution of BLAST.\n",
    "\n",
    "### Installing BLAST\n",
    "I suspect most everyone here is accustomed to using the NCBI BLAST web interface, but it's just another program, whether it's running on an NCBI web server, or on our laptops. First we will install BLAST locally, such that we have ready access to it from anywhere on our computers.\n",
    "\n",
    "#### Windows\n",
    "\n",
    "Windows users have the easiest time installing BLAST.\n",
    "\n",
    "Go to the PythonCourse/ProgramFiles/windows folder. Run the ncbi-blast-2.4.0+-win64.exe file (either by double clicking it in a file browser or invoking it within a shell), and follow the install GUI. It should automatically install all the different BLAST programs, and add them to your path. \n",
    "\n",
    "To verify your install worked, open a fresh terminal a type 'makeblastdb.' The program should give you an error (as you gave it no input), but should display the command line help. We will cover how to use the different BLAST programs below.\n",
    "\n",
    "#### Mac and Linux\n",
    "\n",
    "In the PythonCourse/ProgramFiles directory, there is a *.tar.gz* file containing the BLAST executable for your platform. The *.tar.gz* file extension tells us that the file has been archived (the *.tar* part) and compressed (the *.gz* part). The UNIX command to uncompress and expand the file is:\n",
    "\n",
    "**tar -zxvf [filename]**\n",
    "\n",
    "And we can tell **tar** where to dump the output with the **-C** option, like so:\n",
    "\n",
    "**tar -zxvf [filename] -C [output directory]**\n",
    "\n",
    "Let's use __check_output()__ to do this now. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ncbi-blast-2.2.26+/\n",
      "ncbi-blast-2.2.26+/bin/\n",
      "ncbi-blast-2.2.26+/bin/blastn\n",
      "ncbi-blast-2.2.26+/bin/rpstblastn\n",
      "ncbi-blast-2.2.26+/bin/convert2blastmask\n",
      "ncbi-blast-2.2.26+/bin/blastdbcmd\n",
      "ncbi-blast-2.2.26+/bin/tblastn\n",
      "ncbi-blast-2.2.26+/bin/makeblastdb\n",
      "ncbi-blast-2.2.26+/bin/makeprofiledb\n",
      "ncbi-blast-2.2.26+/bin/segmasker\n",
      "ncbi-blast-2.2.26+/bin/windowmasker\n",
      "ncbi-blast-2.2.26+/bin/update_blastdb.pl\n",
      "ncbi-blast-2.2.26+/bin/blastdb_aliastool\n",
      "ncbi-blast-2.2.26+/bin/rpsblast\n",
      "ncbi-blast-2.2.26+/bin/psiblast\n",
      "ncbi-blast-2.2.26+/bin/dustmasker\n",
      "ncbi-blast-2.2.26+/bin/makembindex\n",
      "ncbi-blast-2.2.26+/bin/blastp\n",
      "ncbi-blast-2.2.26+/bin/tblastx\n",
      "ncbi-blast-2.2.26+/bin/blastx\n",
      "ncbi-blast-2.2.26+/bin/blastdbcheck\n",
      "ncbi-blast-2.2.26+/bin/deltablast\n",
      "ncbi-blast-2.2.26+/bin/blast_formatter\n",
      "ncbi-blast-2.2.26+/bin/legacy_blast.pl\n",
      "ncbi-blast-2.2.26+/doc/\n",
      "ncbi-blast-2.2.26+/doc/README.txt\n",
      "ncbi-blast-2.2.26+/ChangeLog\n",
      "ncbi-blast-2.2.26+/ncbi_package_info\n",
      "ncbi-blast-2.2.26+/README\n",
      "ncbi-blast-2.2.26+/LICENSE\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "# Make sure this is right for you.\n",
    "blast_archive = ('~/Dropbox/2018_winter_python/ProgramFiles/' + \\\n",
    "                 'linux/ncbi-blast-2.2.26+-x64-linux.tar.gz')\n",
    "blast_destination = '~/Dropbox/2018_winter_python/ProgramFiles'\n",
    "\n",
    "command = 'tar -zxvf {} -C {}'.format(blast_archive, blast_destination)\n",
    "\n",
    "# This may take a few seconds\n",
    "print sp.check_output(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Within the BLAST directory we've got a *ChangeLog* file that will tell us what version of BLAST we've got, a *doc* directory with documentation, a *data* directory with the various matrices to use for esimating substituion rates, and then the *bin* directory. *bin* stands for binary, and directories named *bin* are conventionally used to store executable binary files. Try typing **which less** on a terminal and you will see that **less**, just like all your terminal commands (**ls**, **mv**, **cat**, etc), is actually an executable binary, probably in a directory like */usr/bin*.\n",
    "\n",
    "Just like we added our modules to our PYTHONPATH environment variable, we can add executables to our shell's PATH. To do this, in a terminal, type:\n",
    "\n",
    "```bash\n",
    "sudo nano ~/.bashrc\n",
    "```\n",
    "\n",
    "Then add the following lines. __MAKE CERTAIN YOU TYPE THIS EXACTLY CORRECTLY! BREAKING YOUR PATH CAN BE HARD TO FIX__:\n",
    "```bash\n",
    "PATH=$PATH:~/Dropbox/2018_winter_python/ProgramFiles/ncbi-blast-2.2.26+/bin\n",
    "export PATH\n",
    "```\n",
    "\n",
    "Hit CONTROL + x to quit, then type 'y' and hit ENTER to save.\n",
    "\n",
    "Reload your *.bashrc*:\n",
    "\n",
    "```bash\n",
    "source ~/.bashrc\n",
    "```\n",
    "\n",
    "And check if it worked using **which**:\n",
    "\n",
    "```bash\n",
    "which makeblastdb\n",
    "which blastn\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a BLAST Database\n",
    "\n",
    "Now that we have these programs available, let's start using them.\n",
    "\n",
    "Since BLAST is going to compare a query sequence to a database of sequences, collected from say, a genome, or a group of genomes, we first need to create the reference database to BLAST against. Many of these are available for download, but we're going to create our own from the *S. cerevisiae* genome.\n",
    "\n",
    "The yeast genome is available on the NCBI FTP server, though we have included it in your course folder. *yeastn.nt* is the fasta formated genome, while *yeast.aa* is the fasta formated collection of all proteins.\n",
    "\n",
    "To make your own local BLAST database, use the **makeblastdb** program we just installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "\r\n",
      "Building a new DB, current time: 12/23/2017 11:49:29\r\n",
      "New DB name:   C:\\Users\\trahs\\Dropbox\\2018_winter_python\\4.2_Modules_and_System_Calls\\yeast_genome\r\n",
      "New DB title:  yeast.nt\r\n",
      "Sequence type: Nucleotide\r\n",
      "Deleted existing Nucleotide BLAST database named C:\\Users\\trahs\\Dropbox\\2018_winter_python\\4.2_Modules_and_System_Calls\\yeast_genome\r\n",
      "Keep MBits: T\r\n",
      "Maximum file size: 1000000000B\r\n",
      "Adding sequences from FASTA; added 17 sequences in 0.199087 seconds.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "command = 'makeblastdb -in yeast.nt -out yeast_genome -dbtype nucl'\n",
    "\n",
    "print sp.check_output(command,shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have a fully operational BLAST installation and a yeast nucleotide database to search against. We can see that BLAST has created some additional files for our yeast database in our working directory (*.nhr*, *.nin*, and *.nsq*).\n",
    "\n",
    "### Running BLAST!\n",
    "\n",
    "Now let's query the database with __blastn__. We will have to make an input file to serve as our query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTN 2.4.0+\r\n",
      "\r\n",
      "\r\n",
      "Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb\r\n",
      "Miller (2000), \"A greedy algorithm for aligning DNA sequences\", J\r\n",
      "Comput Biol 2000; 7(1-2):203-14.\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "Database: yeast.nt\r\n",
      "           17 sequences; 12,155,026 total letters\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "Query= \r\n",
      "Length=341\r\n",
      "                                                                      Score     E\r\n",
      "Sequences producing significant alignments:                          (Bits)  Value\r\n",
      "\r\n",
      "  gi|6321173|ref|NC_001139.1| Saccharomyces cerevisiae chromosome...  623     8e-179\r\n",
      "  gi|7276232|ref|NC_001137.2| Saccharomyces cerevisiae chromosome...  597     5e-171\r\n",
      "  gi|7839148|ref|NC_001136.2| Saccharomyces cerevisiae chromosome...  592     2e-169\r\n",
      "  gi|6322623|ref|NC_001143.1| Saccharomyces cerevisiae chromosome...  584     4e-167\r\n",
      "  gi|6322016|ref|NC_001141.1| Saccharomyces cerevisiae chromosome...  584     4e-167\r\n",
      "  gi|6319247|ref|NC_001133.1| Saccharomyces cerevisiae chromosome...  584     4e-167\r\n",
      "  gi|6324971|ref|NC_001148.1| Saccharomyces cerevisiae chromosome...  580     5e-166\r\n",
      "  gi|6862570|ref|NC_001140.2| Saccharomyces cerevisiae chromosome...  580     5e-166\r\n",
      "  gi|6324406|ref|NC_001147.1| Saccharomyces cerevisiae chromosome...  579     2e-165\r\n",
      "  gi|6323989|ref|NC_001146.1| Saccharomyces cerevisiae chromosome...  579     2e-165\r\n",
      "  gi|6319354|ref|NC_001134.1| Saccharomyces cerevisiae chromosome...  575     2e-164\r\n",
      "  gi|6322960|ref|NC_001144.1| Saccharomyces cerevisiae chromosome...  568     4e-162\r\n",
      "\r\n",
      "\r\n",
      "> gi|6321173|ref|NC_001139.1| Saccharomyces cerevisiae chromosome \r\n",
      "VII, complete chromosome sequence\r\n",
      "Length=1090936\r\n",
      "\r\n",
      " Score = 623 bits (337),  Expect = 8e-179\r\n",
      " Identities = 340/341 (99%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319421  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  319480\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319481  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  319540\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319541  TTATATATTATATAAAATATGA-ATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  319599\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319600  AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  319659\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319660  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  319719\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  319720  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  319760\r\n",
      "\r\n",
      "\r\n",
      " Score = 592 bits (320),  Expect = 2e-169\r\n",
      " Identities = 335/342 (98%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||| ||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  736316  TGTTGAATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  736257\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  736256  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  736197\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  736196  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  736138\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||| \r\n",
      "Sbjct  736137  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  736078\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  736077  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  736018\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  736017  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  735976\r\n",
      "\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  707528  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  707469\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  707468  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  707409\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  707408  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  707350\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  707349  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  707290\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  707289  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  707230\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  707229  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  707189\r\n",
      "\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  712539  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  712480\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  712479  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  712420\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  712419  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  712361\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  712360  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  712301\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  712300  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  712241\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  712240  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  712200\r\n",
      "\r\n",
      "\r\n",
      " Score = 477 bits (258),  Expect = 7e-135\r\n",
      " Identities = 271/277 (98%), Gaps = 2/277 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  66      AAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATCTTATA  125\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  401888  AAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATCTTATA  401829\r\n",
      "\r\n",
      "Query  126     TATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCTAAACT  185\r\n",
      "               ||||||||||||||||| |||||| |||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  401828  TATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCTAAACT  401770\r\n",
      "\r\n",
      "Query  186     AATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTATCTCA  244\r\n",
      "               |||||||||||||||||||||||||| |||||||||||||||||| ||||||||||||||\r\n",
      "Sbjct  401769  AATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGCCGGGTTTTATCTCA  401710\r\n",
      "\r\n",
      "Query  245     TGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGT  304\r\n",
      "               |||||||||||||| ||||||||||||||||||||||| |||||||||||||||||||||\r\n",
      "Sbjct  401709  TGTTGTTCGTTTTGATATAAGGTTGTTTCATATGTGTTGTATGAACGTTTAGGATGACGT  401650\r\n",
      "\r\n",
      "Query  305     ATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  401649  ATTGTCATACTGACATATCTCATTTTGAGATACAACA  401613\r\n",
      "\r\n",
      "\r\n",
      " Score = 340 bits (184),  Expect = 9e-094\r\n",
      " Identities = 276/319 (87%), Gaps = 12/319 (4%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  34      TCTTGACAATCCTAATATAAACGGTCTTAGGGAAGTAACCAGTTGTCAAAACAGTTTATC  93\r\n",
      "               |||||||||||||||||||||| ||||||||||||||||||||||||||||| |||||||\r\n",
      "Sbjct  845592  TCTTGACAATCCTAATATAAACAGTCTTAGGGAAGTAACCAGTTGTCAAAACGGTTTATC  845533\r\n",
      "\r\n",
      "Query  94      AGATTAATTCACGGAATGTTACTTATCTTATATATTATATAAAATATGATATCATATTAA  153\r\n",
      "               || || |||||||| || ||||||||| ||||||||||||||||||||  |||||| |||\r\n",
      "Sbjct  845532  AGCTTGATTCACGGGATCTTACTTATCCTATATATTATATAAAATATGG-ATCATACTAA  845474\r\n",
      "\r\n",
      "Query  154     GTGGTGGAAGCGCGGAATCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTTG-GT  212\r\n",
      "                ||||||||||||||||||| |||||||||| ||||||||||  | |||||| || | ||\r\n",
      "Sbjct  845473  ATGGTGGAAGCGCGGAATCTTGGATCTAAACCAATTGTTCAGATACTTATACGTTAGAGT  845414\r\n",
      "\r\n",
      "Query  213     AGTTCAGCTAGGGAAGGACGGGTTTTATCT-CATGTTG-T--T-C-GTT--TTGT--TAT  262\r\n",
      "               || | ||||| ||||||  | ||||||| | ||||||| |  | | |||  || |  | |\r\n",
      "Sbjct  845413  AGGTTAGCTAAGGAAGGCAGAGTTTTATTTTCATGTTGCTCATTCTGTTCGTTTTGCTGT  845354\r\n",
      "\r\n",
      "Query  263     AAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGACATAT  322\r\n",
      "               ||||||||||||||||||||||||||||||||||||||| |||||||||||||||  |||\r\n",
      "Sbjct  845353  AAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGATGTATTGTCATACTGAGGTAT  845294\r\n",
      "\r\n",
      "Query  323     CTCATTTTGAGATACAACA  341\r\n",
      "               | |||| ||||||||||||\r\n",
      "Sbjct  845293  CCCATTCTGAGATACAACA  845275\r\n",
      "\r\n",
      "\r\n",
      " Score = 339 bits (183),  Expect = 3e-093\r\n",
      " Identities = 217/232 (94%), Gaps = 7/232 (3%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  111     GTTACTTATCTTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAA  170\r\n",
      "               |||||||||||||||||||||||||||||||| |||||| ||||||||||||||||||||\r\n",
      "Sbjct  405225  GTTACTTATCTTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAA  405283\r\n",
      "\r\n",
      "Query  171     TCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGG  229\r\n",
      "               ||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||\r\n",
      "Sbjct  405284  TCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGG  405343\r\n",
      "\r\n",
      "Query  230     ACGGGTTTTATCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAA  289\r\n",
      "                |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  405344  CCGGGTTTTATCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAA  405403\r\n",
      "\r\n",
      "Query  290     CGTTTAGGATGACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               || || | |  | ||||||||||||||||   ||| |||||||||| |||||\r\n",
      "Sbjct  405404  CGGTT-G-A--A-GTATTGTCATACTGACGCTTCTGATTTTGAGATTCAACA  405450\r\n",
      "\r\n",
      "\r\n",
      "> gi|7276232|ref|NC_001137.2| Saccharomyces cerevisiae chromosome \r\n",
      "V, complete chromosome sequence\r\n",
      "Length=576869\r\n",
      "\r\n",
      " Score = 597 bits (323),  Expect = 5e-171\r\n",
      " Identities = 336/342 (98%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  434627  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  434686\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  434687  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  434746\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  434747  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  434805\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||| |||||||||||||||||||| |||||||||\r\n",
      "Sbjct  434806  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGCCGGGTTTTA  434865\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  434866  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  434925\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  434926  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  434967\r\n",
      "\r\n",
      "\r\n",
      " Score = 592 bits (320),  Expect = 2e-169\r\n",
      " Identities = 335/342 (98%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  249929  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  249988\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  249989  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  250048\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  250049  TTATATATTATATAAAATATGA-ATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  250107\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||| ||| |||||||||||||||||||| |||||||||\r\n",
      "Sbjct  250108  AAACTAATTGTTCAGGCATTTATACGTTTGGGTAGTTCAGCTAGGGAAGGCCGGGTTTTA  250167\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               |||||||||||||||||||||||||||||||||||| |||||||||||||||||||||||\r\n",
      "Sbjct  250168  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATCTGTGTTTTATGAACGTTTAGGAT  250227\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  250228  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  250269\r\n",
      "\r\n",
      "\r\n",
      " Score = 448 bits (242),  Expect = 5e-126\r\n",
      " Identities = 253/258 (98%), Gaps = 2/258 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  492434  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  492493\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  492494  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  492553\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  492554  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  492612\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||| \r\n",
      "Sbjct  492613  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  492672\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTT  257\r\n",
      "               ||||||||||||||||||\r\n",
      "Sbjct  492673  TCTCATGTTGTTCGTTTT  492690\r\n",
      "\r\n",
      "\r\n",
      " Score = 159 bits (86),  Expect = 3e-039\r\n",
      " Identities = 88/89 (99%), Gaps = 0/89 (0%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  253     GTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCAT  312\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  498417  GTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCAT  498476\r\n",
      "\r\n",
      "Query  313     ACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||| ||||||||||||||||||||||\r\n",
      "Sbjct  498477  ACTGACGTATCTCATTTTGAGATACAACA  498505\r\n",
      "\r\n",
      "\r\n",
      "> gi|7839148|ref|NC_001136.2| Saccharomyces cerevisiae chromosome \r\n",
      "IV, complete chromosome sequence\r\n",
      "Length=1531929\r\n",
      "\r\n",
      " Score = 592 bits (320),  Expect = 2e-169\r\n",
      " Identities = 335/342 (98%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||||||||||||||||||||||||||||||||| |||||||||||||||||||| ||||\r\n",
      "Sbjct  945946  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTTTTGACAATCCTAATATAAACAGTCT  946005\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  946006  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  946065\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  946066  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  946124\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||| \r\n",
      "Sbjct  946125  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  946184\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  946185  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  946244\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  946245  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  946286\r\n",
      "\r\n",
      "\r\n",
      " Score = 586 bits (317),  Expect = 1e-167\r\n",
      " Identities = 334/342 (98%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  668429  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  668370\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  668369  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  668310\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  668309  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  668251\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATAC-TTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               |||||||||||||| |||||||||| ||| |||||||||||||||||||| || ||||||\r\n",
      "Sbjct  668250  AAACTAATTGTTCAAGCATTTATACGTTTGGGTAGTTCAGCTAGGGAAGGCCGTGTTTTA  668191\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  668190  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  668131\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  668130  GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  668089\r\n",
      "\r\n",
      "\r\n",
      " Score = 564 bits (305),  Expect = 5e-161\r\n",
      " Identities = 330/342 (96%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1        TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "                |||||||||||||||||||||||||||||||||||||||||||||| |||||||| ||||\r\n",
      "Sbjct  1023468  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTGATATAAACAGTCT  1023409\r\n",
      "\r\n",
      "Query  61       TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "                ||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  1023408  TAGGGAAGTAACCAGTTGTCAAAACGGTTTATCAGATTAATTCACGGAATGTTACTTATC  1023349\r\n",
      "\r\n",
      "Query  121      TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "                |||| ||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  1023348  TTATGTATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  1023290\r\n",
      "\r\n",
      "Query  181      AAACTAATTGTTCAGGCATTTATAC-TTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "                ||||| |||||||| |||||||||| ||| |||||||||||||||||||| || ||||||\r\n",
      "Sbjct  1023289  AAACTGATTGTTCAAGCATTTATACGTTTGGGTAGTTCAGCTAGGGAAGGCCGTGTTTTA  1023230\r\n",
      "\r\n",
      "Query  240      TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "                ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  1023229  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  1023170\r\n",
      "\r\n",
      "Query  300      GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "                ||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  1023169  GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  1023128\r\n",
      "\r\n",
      "\r\n",
      " Score = 429 bits (232),  Expect = 2e-120\r\n",
      " Identities = 253/263 (96%), Gaps = 2/263 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  645233  TGTTGTATTACGGGCTCGAGTAATACCGCAGTGTCTTGACAATCCTAATATAAACAGTCT  645292\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||||||||||||| |||||||\r\n",
      "Sbjct  645293  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTCACTTATC  645352\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  645353  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  645411\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               |||||||||||||| |||||||||| ||| |||||||||||||||||||| || ||||||\r\n",
      "Sbjct  645412  AAACTAATTGTTCAAGCATTTATACGTTTGGGTAGTTCAGCTAGGGAAGGCCGTGTTTTA  645471\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTAT  262\r\n",
      "               |||||||||||||||||||||||\r\n",
      "Sbjct  645472  TCTCATGTTGTTCGTTTTGTTAT  645494\r\n",
      "\r\n",
      "\r\n",
      " Score = 268 bits (145),  Expect = 4e-072\r\n",
      " Identities = 170/182 (93%), Gaps = 1/182 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  4        TGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCTTAG  63\r\n",
      "                |||| ||||||||| |||||||||||||||||| |||||||||||||||||| |||||||\r\n",
      "Sbjct  1175126  TGTACTACGGGCTCAAGTAATACCGGAGTGTCTCGACAATCCTAATATAAACAGTCTTAG  1175185\r\n",
      "\r\n",
      "Query  64       GGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATCTTA  123\r\n",
      "                |||||||||||||||||||||| |||||||||||||||||||||||||||||||||| ||\r\n",
      "Sbjct  1175186  GGAAGTAACCAGTTGTCAAAACGGTTTATCAGATTAATTCACGGAATGTTACTTATCCTA  1175245\r\n",
      "\r\n",
      "Query  124      TATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCTAAA  183\r\n",
      "                ||||||||||||||||||  |||||| ||| |||||||||||||||| || |||||||||\r\n",
      "Sbjct  1175246  TATATTATATAAAATATGG-ATCATACTAAATGGTGGAAGCGCGGAACCTTGGATCTAAA  1175304\r\n",
      "\r\n",
      "Query  184      CT  185\r\n",
      "                ||\r\n",
      "Sbjct  1175305  CT  1175306\r\n",
      "\r\n",
      "\r\n",
      " Score = 156 bits (84),  Expect = 4e-038\r\n",
      " Identities = 84/84 (100%), Gaps = 0/84 (0%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  258     GTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGA  317\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  651413  GTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGA  651472\r\n",
      "\r\n",
      "Query  318     CATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||||||||\r\n",
      "Sbjct  651473  CATATCTCATTTTGAGATACAACA  651496\r\n",
      "\r\n",
      "\r\n",
      "> gi|6322623|ref|NC_001143.1| Saccharomyces cerevisiae chromosome \r\n",
      "XI, complete chromosome sequence\r\n",
      "Length=666445\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  202287  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  202346\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  202347  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  202406\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  202407  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  202465\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  202466  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  202525\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  202526  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  202585\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  202586  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  202626\r\n",
      "\r\n",
      "\r\n",
      "> gi|6322016|ref|NC_001141.1| Saccharomyces cerevisiae chromosome \r\n",
      "IX, complete chromosome sequence\r\n",
      "Length=439885\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  205556  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  205497\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  205496  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  205437\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  205436  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  205378\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  205377  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  205318\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  205317  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  205258\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  205257  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  205217\r\n",
      "\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  210644  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  210585\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  210584  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  210525\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  210524  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  210466\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  210465  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  210406\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  210405  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  210346\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  210345  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  210305\r\n",
      "\r\n",
      "\r\n",
      " Score = 472 bits (255),  Expect = 3e-133\r\n",
      " Identities = 316/345 (92%), Gaps = 5/345 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTA--A-TATAAACGG  57\r\n",
      "               ||||||||||||||||||||||||||||| |||||||||||||| ||  | ||| ||| |\r\n",
      "Sbjct  324732  TGTTGTATTACGGGCTCGAGTAATACCGGGGTGTCTTGACAATCTTATTATTATTAACAG  324673\r\n",
      "\r\n",
      "Query  58      TCTTAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTT  117\r\n",
      "               ||||||||||||||| || ||||||||| | ||||||| ||||||||||| |||||||||\r\n",
      "Sbjct  324672  TCTTAGGGAAGTAACGAGCTGTCAAAACGGATTATCAGCTTAATTCACGGGATGTTACTT  324613\r\n",
      "\r\n",
      "Query  118     ATCTTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGA  177\r\n",
      "               |||||||||||| |||| ||||||| |||||| ||| ||||| ||||||||||||| |||\r\n",
      "Sbjct  324612  ATCTTATATATTGTATAGAATATGA-ATCATACTAAATGGTGAAAGCGCGGAATCTTGGA  324554\r\n",
      "\r\n",
      "Query  178     TCTAAACTAATTGTTCAGGCATTTATACTTTTGGT-AGTTCAGCTAGGGAAGGACGGGTT  236\r\n",
      "               |||||||||||||||||||||||||||| |||||| ||||||||||||| ||| ||||||\r\n",
      "Sbjct  324553  TCTAAACTAATTGTTCAGGCATTTATACGTTTGGTTAGTTCAGCTAGGGGAGGCCGGGTT  324494\r\n",
      "\r\n",
      "Query  237     TTATCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAG  296\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||| | ||||||||\r\n",
      "Sbjct  324493  TTATCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTACGGACGTTTAG  324434\r\n",
      "\r\n",
      "Query  297     GATGACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||| |||| |||| |||||||||||||||||\r\n",
      "Sbjct  324433  GATGACGTATTGTCATATTGACGTATCGCATTTTGAGATACAACA  324389\r\n",
      "\r\n",
      "\r\n",
      "> gi|6319247|ref|NC_001133.1| Saccharomyces cerevisiae chromosome \r\n",
      "I, complete chromosome sequence\r\n",
      "Length=230203\r\n",
      "\r\n",
      " Score = 584 bits (316),  Expect = 4e-167\r\n",
      " Identities = 333/341 (98%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  182949  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  182890\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  182889  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  182830\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  182829  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  182771\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  182770  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  182711\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  182710  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  182651\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  182650  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  182610\r\n",
      "\r\n",
      "\r\n",
      "> gi|6324971|ref|NC_001148.1| Saccharomyces cerevisiae chromosome \r\n",
      "XVI, complete chromosome sequence\r\n",
      "Length=948061\r\n",
      "\r\n",
      " Score = 580 bits (314),  Expect = 5e-166\r\n",
      " Identities = 333/342 (97%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  769656  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  769597\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  769596  TAGGGAAGTAACCAGTTGTTAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  769537\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               ||||||| |||||||||||||| |||||| |||||||||||| |||||||||||||||||\r\n",
      "Sbjct  769536  TTATATACTATATAAAATATGA-ATCATACTAAGTGGTGGAAACGCGGAATCTCGGATCT  769478\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||| \r\n",
      "Sbjct  769477  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  769418\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||||||||||||||| |||||\r\n",
      "Sbjct  769417  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGAT  769358\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  769357  GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  769316\r\n",
      "\r\n",
      "\r\n",
      " Score = 579 bits (313),  Expect = 2e-165\r\n",
      " Identities = 326/332 (98%), Gaps = 2/332 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  856880  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  856821\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  856820  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  856761\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  856760  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  856702\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||| \r\n",
      "Sbjct  856701  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  856642\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||||||||||||||| |||||\r\n",
      "Sbjct  856641  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGAT  856582\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTG  331\r\n",
      "               ||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  856581  GACGTATTGTCATACTGACATATCTCATTTTG  856550\r\n",
      "\r\n",
      "\r\n",
      " Score = 520 bits (281),  Expect = 1e-147\r\n",
      " Identities = 322/342 (94%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  880721  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  880662\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||| ||||||||| || |||||||| ||||||||||||\r\n",
      "Sbjct  880661  TAGGGAAGTAACCAGTTGTCAAAACTGTTTATCAGCTTGATTCACGGGATGTTACTTATC  880602\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "                ||||||||||||||||||||| |||||| ||| |||||||||||||||| |||||||||\r\n",
      "Sbjct  880601  CTATATATTATATAAAATATGA-ATCATACTAAATGGTGGAAGCGCGGAAACTCGGATCT  880543\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTG-GTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||| |||| || |||||||||||| ||| |||||||||\r\n",
      "Sbjct  880542  AAACTAATTGTTCAGGCATTTATACGTTTGAGTTGTTCAGCTAGGGGAGGCCGGGTTTTA  880483\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "                |||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  880482  CCTCATGTTGTTCGTTTTGTTATAAAGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  880423\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               || |||||||||||||||| |||| |||||||||||||||||\r\n",
      "Sbjct  880422  GAAGTATTGTCATACTGACGTATCCCATTTTGAGATACAACA  880381\r\n",
      "\r\n",
      "\r\n",
      "> gi|6862570|ref|NC_001140.2| Saccharomyces cerevisiae chromosome \r\n",
      "VIII, complete chromosome sequence\r\n",
      "Length=562639\r\n",
      "\r\n",
      " Score = 580 bits (314),  Expect = 5e-166\r\n",
      " Identities = 333/342 (97%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||| | ||| ||||\r\n",
      "Sbjct  146322  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATCTGAACAGTCT  146381\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  146382  TAGGGAAGTAACCAGTTGTCAAAACGGTTTATCAGATTAATTCACGGAATGTTACTTATC  146441\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  146442  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  146500\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||| \r\n",
      "Sbjct  146501  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  146560\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  146561  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  146620\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  146621  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  146662\r\n",
      "\r\n",
      "\r\n",
      " Score = 248 bits (134),  Expect = 6e-066\r\n",
      " Identities = 146/151 (97%), Gaps = 3/151 (2%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  1      TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "              |||||||||||||||||||||||||||||||||||||||||||||||||| |||| ||||\r\n",
      "Sbjct  85383  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATACAAACAGTCT  85442\r\n",
      "\r\n",
      "Query  61     TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "              ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  85443  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  85502\r\n",
      "\r\n",
      "Query  121    T--TATATATTATATAAAATATGATATCATA  149\r\n",
      "              |  ||||||||||||||||||||| ||||||\r\n",
      "Sbjct  85503  TTATATATATTATATAAAATATGA-ATCATA  85532\r\n",
      "\r\n",
      "\r\n",
      "> gi|6324406|ref|NC_001147.1| Saccharomyces cerevisiae chromosome \r\n",
      "XV, complete chromosome sequence\r\n",
      "Length=1091283\r\n",
      "\r\n",
      " Score = 579 bits (313),  Expect = 2e-165\r\n",
      " Identities = 333/342 (97%), Gaps = 3/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| | ||||\r\n",
      "Sbjct  854610  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAA-CAGTCT  854552\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  854551  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  854492\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  854491  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  854433\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||| ||||||||||||||| \r\n",
      "Sbjct  854432  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAAGGAAGGACGGGTTTTG  854373\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               |||||||||||||| ||||||||||||||||||||||||||||||||||||||| |||||\r\n",
      "Sbjct  854372  TCTCATGTTGTTCGCTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGAT  854313\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  854312  GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  854271\r\n",
      "\r\n",
      "\r\n",
      " Score = 300 bits (162),  Expect = 2e-081\r\n",
      " Identities = 266/314 (85%), Gaps = 15/314 (5%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  39      ACAATCCTAATATAAACGGTCTTAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATT  98\r\n",
      "               ||||||||||||||| | ||||| |||||| |||||| ||||||||| | ||| ||| ||\r\n",
      "Sbjct  969293  ACAATCCTAATATAAGCAGTCTTGGGGAAGCAACCAGCTGTCAAAACGGATTACCAGCTT  969234\r\n",
      "\r\n",
      "Query  99      AATTCACGGAATGTTACTTATCTTATATATTATATAAAATATG-ATATCATATTAAGTGG  157\r\n",
      "                |||| |||  ||||||||||| |||||||||||||||||||| || | ||| ||| || \r\n",
      "Sbjct  969233  GATTCGCGGGCTGTTACTTATCCTATATATTATATAAAATATGGAT-T-ATACTAAATGA  969176\r\n",
      "\r\n",
      "Query  158     TGGAAGCGCGGAATCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTT  216\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||||  ||| ||||| |||||\r\n",
      "Sbjct  969175  TGGAAGCGCGGAATCTCGGATCTAAACTAATTGTTCAGGCATTCGTACGTTTGGGTAGTT  969116\r\n",
      "\r\n",
      "Query  217     CAGCTAGGGAAGGACGG-GTTTTATCTCATGTTG----TTC-GT----TTTGTTATAAGG  266\r\n",
      "               |||||| || | | ||| |||| |||||||||||    ||  ||    |||||||||| |\r\n",
      "Sbjct  969115  CAGCTAAGGGACG-CGGAGTTTCATCTCATGTTGCCCATTTTGTTCGCTTTGTTATAAAG  969057\r\n",
      "\r\n",
      "Query  267     TTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGACATATCTCA  326\r\n",
      "               ||||||||||||||||||| |||||||||| |||| |||||||||||||||   ||||||\r\n",
      "Sbjct  969056  TTGTTTCATATGTGTTTTACGAACGTTTAGAATGATGTATTGTCATACTGAGGGATCTCA  968997\r\n",
      "\r\n",
      "Query  327     TTTTGAGATACAAC  340\r\n",
      "               || |||||||||||\r\n",
      "Sbjct  968996  TTCTGAGATACAAC  968983\r\n",
      "\r\n",
      "\r\n",
      " Score = 270 bits (146),  Expect = 1e-072\r\n",
      " Identities = 284/347 (82%), Gaps = 23/347 (7%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||||||||| |||||||||||||||||| |||||| ||||||||||||||||   ||||\r\n",
      "Sbjct  227036  TGTTGTATTA-GGGCTCGAGTAATACCGGGGTGTCTAGACAATCCTAATATAAGGAGTCT  226978\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               | ||||||||| |||||| |||||| | ||||||| || ||||| || ||||| ||||| \r\n",
      "Sbjct  226977  TTGGGAAGTAATCAGTTGCCAAAACGGATTATCAGCTTGATTCATGGGATGTTGCTTATT  226918\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "                ||||||| ||||||||||||| || | | ||| || || ||||| |||||||| |||||\r\n",
      "Sbjct  226917  CTATATATCATATAAAATATGA-ATTACACTAAATGATGAAAGCGAGGAATCTCAGATCT  226859\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATAC-TTTTGGTAGTTCAGCTAGGGAAGGACGG-GTTTT  238\r\n",
      "               |||||||||||||||| |||||||| ||| ||| ||||| ||| || ||| ||| |||||\r\n",
      "Sbjct  226858  AAACTAATTGTTCAGGGATTTATACGTTTGGGTGGTTCAACTACGGGAGG-CGGAGTTTT  226800\r\n",
      "\r\n",
      "Query  239     ATCTCATGTTG---------TTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAA  289\r\n",
      "               |||||||||||         ||||||||||||||||||||||||||||| |  || | | \r\n",
      "Sbjct  226799  ATCTCATGTTGCTCATTCTATTCGTTTTGTTATAAGGTTGTTTCATATGAGCGTT-T-A-  226743\r\n",
      "\r\n",
      "Query  290     CGTTTAGGATGACGTATTGTCATACTGACATAT-CTCATTTTGAGAT  335\r\n",
      "               ||   | ||||| |||||||||||||||  ||| || ||  ||||||\r\n",
      "Sbjct  226742  CG---A-GATGATGTATTGTCATACTGAGGTATGCT-ATCCTGAGAT  226701\r\n",
      "\r\n",
      "\r\n",
      " Score = 263 bits (142),  Expect = 2e-070\r\n",
      " Identities = 287/353 (81%), Gaps = 26/353 (7%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||||| |||||||  ||||||||||||| ||||||||| ||||| ||||||||| ||| \r\n",
      "Sbjct  354466  TGTTGTTTTACGGGTACGAGTAATACCGGGGTGTCTTGAAAATCCCAATATAAACAGTC-  354408\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "                | |||||||||||||||||| ||| | ||||||| ||  |||| || ||||||||||||\r\n",
      "Sbjct  354407  AATGGAAGTAACCAGTTGTCAGAACGGATTATCAGCTT--TTCATGGGATGTTACTTATC  354350\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "                ||||||||||||||||||||  |||||||||| |||||||||||| ||| |||||||||\r\n",
      "Sbjct  354349  CTATATATTATATAAAATATG-GATCATATTAATTGGTGGAAGCGCCGAACCTCGGATCT  354291\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATAC-TTTTGGTAGTTCAGCTAGGGAAGGACG--GGT-T  236\r\n",
      "               ||| |||||| |||||| | ||||| ||| ||| ||| ||||| || ||| ||  | | |\r\n",
      "Sbjct  354290  AAATTAATTGCTCAGGCGTCTATACGTTTGGGTGGTTAAGCTAAGGGAGG-CGAAGTTAT  354232\r\n",
      "\r\n",
      "Query  237     -T-T-A---T-CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAA  289\r\n",
      "                | | |   | |||||  |||||||||||||||||||||||||||||||      |  | \r\n",
      "Sbjct  354231  ATGTCATGTTGCTCATTCTGTTCGTTTTGTTATAAGGTTGTTTCATATG------A--A-  354181\r\n",
      "\r\n",
      "Query  290     CG-TTTAGGATGACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               || |||||||||| |||||||||||||||| | |  || | ||||| ||||||\r\n",
      "Sbjct  354180  CGTTTTAGGATGATGTATTGTCATACTGACGTGTACCACTCTGAGAAACAACA  354128\r\n",
      "\r\n",
      "\r\n",
      "> gi|6323989|ref|NC_001146.1| Saccharomyces cerevisiae chromosome \r\n",
      "XIV, complete chromosome sequence\r\n",
      "Length=784328\r\n",
      "\r\n",
      " Score = 579 bits (313),  Expect = 2e-165\r\n",
      " Identities = 332/341 (97%), Gaps = 1/341 (0%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  547073  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  547014\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  547013  TAGGGAAGTAACCAGTTGTCAAAACGGTTTATCAGATTAATTCACGGAATGTTACTTATC  546954\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  546953  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  546895\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGGTAGTTCAGCTAGGGAAGGACGGGTTTTAT  240\r\n",
      "               |||||||||||||||||||||||||||| |||||||||||||||||||| |||||||| |\r\n",
      "Sbjct  546894  AAACTAATTGTTCAGGCATTTATACTTTGGGTAGTTCAGCTAGGGAAGGTCGGGTTTTGT  546835\r\n",
      "\r\n",
      "Query  241     CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATG  300\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||| ||| ||\r\n",
      "Sbjct  546834  CTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGGTG  546775\r\n",
      "\r\n",
      "Query  301     ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  546774  ACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  546734\r\n",
      "\r\n",
      "\r\n",
      " Score = 564 bits (305),  Expect = 5e-161\r\n",
      " Identities = 330/342 (96%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  726570  TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  726511\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  726510  TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  726451\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||  ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  726450  TTATATATTATATAAAATATGA-ATCATGCTAAGTGGTGGAAGCGCGGAATCTCGGATCT  726392\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||| \r\n",
      "Sbjct  726391  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  726332\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  726331  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  726272\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "                 |  | |||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  726271  CTCAAAATGTCATACTGACGTATCTCATTTTGAGATACAACA  726230\r\n",
      "\r\n",
      "\r\n",
      "> gi|6319354|ref|NC_001134.1| Saccharomyces cerevisiae chromosome \r\n",
      "II, complete chromosome sequence\r\n",
      "Length=813140\r\n",
      "\r\n",
      " Score = 575 bits (311),  Expect = 2e-164\r\n",
      " Identities = 332/342 (97%), Gaps = 2/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               ||||||||||||||||||||||||| ||||||||||||||||||||||||| ||| ||||\r\n",
      "Sbjct  198017  TGTTGTATTACGGGCTCGAGTAATATCGGAGTGTCTTGACAATCCTAATATGAACAGTCT  197958\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  197957  TAGGGAAGTAACCAGTTGTCAAAACGGTTTATCAGATTAATTCACGGAATGTTACTTATC  197898\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  197897  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  197839\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| |||||||||||||||||| |||||||||\r\n",
      "Sbjct  197838  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGCCGGGTTTTA  197779\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               |||||||||||||||||||||||||||||||||||||||||||||||||||||| |||||\r\n",
      "Sbjct  197778  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTCAGGAT  197719\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  197718  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  197677\r\n",
      "\r\n",
      "\r\n",
      "> gi|6322960|ref|NC_001144.1| Saccharomyces cerevisiae chromosome \r\n",
      "XII, complete chromosome sequence\r\n",
      "Length=1078172\r\n",
      "\r\n",
      " Score = 568 bits (307),  Expect = 4e-162\r\n",
      " Identities = 331/342 (97%), Gaps = 3/342 (1%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||| |||||||||||||||||||||||||||||||||||||||||||||||||| ||||\r\n",
      "Sbjct  374339  TGTTTTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACAGTCT  374280\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               ||||||||| ||||||||||||||| |||||||||||||| ||||||| |||||||||||\r\n",
      "Sbjct  374279  TAGGGAAGTTACCAGTTGTCAAAACGGTTTATCAGATTAACTCACGGA-TGTTACTTATC  374221\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCT  180\r\n",
      "               |||||||||||||||||||||| |||||| ||||||||||||||||||||||||||||||\r\n",
      "Sbjct  374220  TTATATATTATATAAAATATGA-ATCATACTAAGTGGTGGAAGCGCGGAATCTCGGATCT  374162\r\n",
      "\r\n",
      "Query  181     AAACTAATTGTTCAGGCATTTATACTTTTGG-TAGTTCAGCTAGGGAAGGACGGGTTTTA  239\r\n",
      "               ||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||| \r\n",
      "Sbjct  374161  AAACTAATTGTTCAGGCATTTATACTTTTGGGTAGTTCAGCTAGGGAAGGACGGGTTTTG  374102\r\n",
      "\r\n",
      "Query  240     TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  299\r\n",
      "               ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "Sbjct  374101  TCTCATGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGAT  374042\r\n",
      "\r\n",
      "Query  300     GACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               ||||||||||||||||||| ||||||||||||||||||||||\r\n",
      "Sbjct  374041  GACGTATTGTCATACTGACGTATCTCATTTTGAGATACAACA  374000\r\n",
      "\r\n",
      "\r\n",
      " Score = 370 bits (200),  Expect = 1e-102\r\n",
      " Identities = 291/335 (87%), Gaps = 5/335 (1%)\r\n",
      " Strand=Plus/Plus\r\n",
      "\r\n",
      "Query  6       TATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCTTAGGG  65\r\n",
      "               ||||||| || | | |||||| || |||| |||||||||||||||||||| ||||| | |\r\n",
      "Sbjct  168047  TATTACGAGCCCAAATAATACTGGGGTGTTTTGACAATCCTAATATAAACAGTCTTCGAG  168106\r\n",
      "\r\n",
      "Query  66      AAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATCTTATA  125\r\n",
      "               |||||||||||||||||||| | ||||||| || ||||| |||||||||  | ||   ||\r\n",
      "Sbjct  168107  AAGTAACCAGTTGTCAAAACGGATTATCAGCTTGATTCAAGGAATGTTAACTGTC--CTA  168164\r\n",
      "\r\n",
      "Query  126     TATTATATAAAATATGATATCATATTAAGTGGTGGAAGCGCGGAATCTCGGATCTAAACT  185\r\n",
      "               ||| ||||||||||||| |||||| ||| |||||||||| ||| |||| |||||||||||\r\n",
      "Sbjct  168165  TATCATATAAAATATGA-ATCATACTAAATGGTGGAAGCACGG-ATCTTGGATCTAAACT  168222\r\n",
      "\r\n",
      "Query  186     AATTGTTCAGGCATTTATACTTTT-GGTAGTTCAGCTAGGGAAGGACGGGTTTTATCTCA  244\r\n",
      "               |||||||||||||||||||  ||| |||||||  |||| || ||| ||||||||||||||\r\n",
      "Sbjct  168223  AATTGTTCAGGCATTTATATGTTTGGGTAGTTTGGCTAAGGGAGGCCGGGTTTTATCTCA  168282\r\n",
      "\r\n",
      "Query  245     TGTTGTTCGTTTTGTTATAAGGTTGTTTCATATGTGTTTTATGAACGTTTAGGATGACGT  304\r\n",
      "               ||||||||||||||||||| | ||||||||||||| ||||||||||||||||||| | ||\r\n",
      "Sbjct  168283  TGTTGTTCGTTTTGTTATAGGATTGTTTCATATGTATTTTATGAACGTTTAGGATAATGT  168342\r\n",
      "\r\n",
      "Query  305     ATTGTCATACTGACATATCTCATTTTGAGATACAA  339\r\n",
      "               |||||||  ||||| ||||||||||||||||||||\r\n",
      "Sbjct  168343  ATTGTCAGGCTGACGTATCTCATTTTGAGATACAA  168377\r\n",
      "\r\n",
      "\r\n",
      " Score = 327 bits (177),  Expect = 7e-090\r\n",
      " Identities = 298/353 (84%), Gaps = 22/353 (6%)\r\n",
      " Strand=Plus/Minus\r\n",
      "\r\n",
      "Query  1       TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAACGGTCT  60\r\n",
      "               |||||||||||||||||||| |||||||| ||  ||||||||||||||||||||| ||||\r\n",
      "Sbjct  657364  TGTTGTATTACGGGCTCGAGCAATACCGGGGTCCCTTGACAATCCTAATATAAACAGTCT  657305\r\n",
      "\r\n",
      "Query  61      TAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGAATGTTACTTATC  120\r\n",
      "               | |||||||||| ||||| |||||  | |||||||     || |||||||||||||||||\r\n",
      "Sbjct  657304  TTGGGAAGTAACGAGTTGCCAAAAGGGATTATCAG----CTT-ACGGAATGTTACTTATC  657250\r\n",
      "\r\n",
      "Query  121     TTATATATTATATAAAATATGATATCATA-TTAAGTGGTGGAAGCGCGGAATCTCGGATC  179\r\n",
      "               |||||||||||||||||||||  |||||| || | |||||||||||||||||||||||||\r\n",
      "Sbjct  657249  TTATATATTATATAAAATATGG-ATCATACTTGA-TGGTGGAAGCGCGGAATCTCGGATC  657192\r\n",
      "\r\n",
      "Query  180     TAAACTAATTGTTCAGGCATTTATACTTTTGGTA-GTTCAGCTAGGGAAGGACGG-GTTT  237\r\n",
      "               |||||||||| ||||||| ||||||| ||||| | |||||| || ||| || ||| ||||\r\n",
      "Sbjct  657191  TAAACTAATTTTTCAGGCGTTTATACGTTTGGAAAGTTCAGTTAAGGA-GG-CGGAGTTT  657134\r\n",
      "\r\n",
      "Query  238     TATCTCATGTTG-T--T-C-GTT--TTGT--TATAAGGTTGTTTCATATGTGTTTTATGA  288\r\n",
      "                ||||||||||| |  | | |||  || |  ||||||||||||||||||||  |||| ||\r\n",
      "Sbjct  657133  CATCTCATGTTGCTCATTCTGTTCGTTTTGCTATAAGGTTGTTTCATATGTAATTTAAGA  657074\r\n",
      "\r\n",
      "Query  289     ACGTTTAGGATGACGTATTGTCATACTGACATATCTCATTTTGAGATACAACA  341\r\n",
      "               |||||||||||||  |||| |||||||||| || | |||| ||| | ||||||\r\n",
      "Sbjct  657073  ACGTTTAGGATGAT-TATTATCATACTGACGTACCCCATTCTGAAACACAACA  657022\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "Lambda      K        H\r\n",
      "    1.33    0.621     1.12 \r\n",
      "\r\n",
      "Gapped\r\n",
      "Lambda      K        H\r\n",
      "    1.28    0.460    0.850 \r\n",
      "\r\n",
      "Effective search space used: 3877333988\r\n",
      "\r\n",
      "\r\n",
      "  Database: yeast.nt\r\n",
      "    Posted date:  Dec 23, 2017  11:49 AM\r\n",
      "  Number of letters in database: 12,155,026\r\n",
      "  Number of sequences in database:  17\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "Matrix: blastn matrix 1 -2\r\n",
      "Gap Penalties: Existence: 0, Extension: 2.5\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "testseq = 'TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAA' \\\n",
    "          'CGGTCTTAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGA' \\\n",
    "          'ATGTTACTTATCTTATATATTATATAAAATATGAtATCATATTAAGTGGTGGAA' \\\n",
    "          'GCGCGGAATCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTTGGTAGTT' \\\n",
    "          'CAGCTAGGGAAGGACGGGTTTTATCTCATGTTGTTCGTTTTGTTATAAGGTTGT' \\\n",
    "          'TTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGACATATCT' \\\n",
    "          'CATTTTGAGATACAACA'\n",
    "\n",
    "def blastn(querySeq):\n",
    "    fh = open('query.seq', 'w')\n",
    "    fh.write(testseq)\n",
    "    fh.close()\n",
    "    \n",
    "    command = 'blastn -db yeast_genome -query query.seq'\n",
    "    out = sp.check_output(command, shell=True)\n",
    "    return out\n",
    "\n",
    "print blastn(testseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can imagine how you use this function in a loop to query as list of sequences, or modify the function to vary the reference database, or even vary some of the BLAST settings (for example, the minimum E-value to report).\n",
    "\n",
    "We have caught the output in Python, which might be helpful when we want to systematically sort through or analyze the results of each BLAST hit. In other circumstances, however, we might want to write the results directly to an output file. Let's try that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "testseq = 'TGTTGTATTACGGGCTCGAGTAATACCGGAGTGTCTTGACAATCCTAATATAAA' \\\n",
    "          'CGGTCTTAGGGAAGTAACCAGTTGTCAAAACAGTTTATCAGATTAATTCACGGA' \\\n",
    "          'ATGTTACTTATCTTATATATTATATAAAATATGAtATCATATTAAGTGGTGGAA' \\\n",
    "          'GCGCGGAATCTCGGATCTAAACTAATTGTTCAGGCATTTATACTTTTGGTAGTT' \\\n",
    "          'CAGCTAGGGAAGGACGGGTTTTATCTCATGTTGTTCGTTTTGTTATAAGGTTGT' \\\n",
    "          'TTCATATGTGTTTTATGAACGTTTAGGATGACGTATTGTCATACTGACATATCT' \\\n",
    "          'CATTTTGAGATACAACA'\n",
    "\n",
    "def blastn(querySeq, outfile):\n",
    "    fh = open('query.seq', 'w')\n",
    "    fh.write(testseq)\n",
    "    fh.close()\n",
    "    \n",
    "    command = 'blastn -db yeast_genome -query query.seq > ' + outfile\n",
    "    sp.check_output(command, shell=True)\n",
    "\n",
    "blastn(testseq, 'blastn_output.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We simply included a second argument to our function, the name of the output file, and used the redirection symbol (__>__) in our shell command to send the output to the file. Note that this function doesn't return anything.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n",
    "\n",
    "__1) Math Overload__\n",
    "\n",
    "Use the **math** module to compute the following:\n",
    "\n",
    "* $4^8$\n",
    "* $log_{10}(3)$\n",
    "* $e^3$\n",
    "* $cos(\\pi)$\n",
    "* $ln(e^3)$\n",
    "\n",
    "__2. Making Directories__\n",
    "\n",
    "Bioinformatics projects require an organized system of files and directories. Work using the **subprocess** module to create a mock file system. Use the subprocess module to do the following (all within a Jupyter Notebook)\n",
    "\n",
    "* (A) Create a *Mock_Project* directory within your /2017_Winter_Python/4.2_Modules_Numpy_Scipy folder\n",
    "* (B) Create *data*, *processed_data*, *scripts*, and *plots* directories.\n",
    "* (C) Print a list of the files in the *Mock_Project* directory\n",
    "* (D) Save the list of all files in the Mock_Project directory to a file called 'directory.txt'\n",
    "\n",
    "\n",
    "__3. BLAST a Fasta File__\n",
    "\n",
    "Write a function that takes the name of a fasta file as input, and prints the results of a BLAST search of every sequence in the file. Use the __blastn__ option __-outfmt 6__ to give the output in summary tabular form. Test this function with the file *test.fasta* in this directory.\n",
    "\n",
    "Feel free to modify BLAST code from the above lecture, and to re-use your Fasta parser from this morning's solutions. Bonus: Allow an option to save the output to a file.\n",
    "\n",
    "\n",
    "__4. Your own BLAST module__\n",
    "\n",
    "Turn the code you wrote above into your own BLAST module, which should include functions to BLAST a specific sequence, as well as to BLAST all sequences in a given Fasta file.\n",
    "\n",
    "Save these funcitons to a file, and add the folder you saved your module in to your PYTHONPATH by modifying your **~/.bashrc** file.\n",
    "\n",
    "Finally, try importing and using your new Python BLAST module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
