{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "##Exercise 1.a##\n",
    "\n",
    "#!/usr/bin/env python\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Function to parse and plot the data\n",
    "def barFPKMs(tableFile, gene):\n",
    "    # Grab the header\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    \n",
    "    # Grab the data for our gene\n",
    "    # The gene names are in the first column\n",
    "    for line in fh:\n",
    "        if line.startswith(gene):\n",
    "            data = line.strip().split()[1:]\n",
    "            data = map(float, data) # Convert all data to a float\n",
    "            break\n",
    "    else:\n",
    "        print \"Gene not found\"\n",
    "        return 0\n",
    "    \n",
    "    # Make a plot\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(1, 1, 1)\n",
    "    \n",
    "    # Matplotlib assumes you have numerical data for both axes\n",
    "    # But for the X axis, we have categorical data\n",
    "    # So we need to plug in a range of numbers to place the bars at\n",
    "    x = range(len(data))\n",
    "    \n",
    "    # Bars will appear at 0, 1, 2, etc\n",
    "    # So set the x-limits to include these values\n",
    "    ax.set_xlim(-1, len(data))\n",
    "    \n",
    "    ax.bar(x, data, align='center')\n",
    "    \n",
    "    # Now we change the labels for our categorical data\n",
    "    labels = [''] + header[1:] # We don't need to label 'Gene' at -1\n",
    "    ax.set_xticklabels(labels)\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "barFPKMs('FPKM_table.tab', 'MAK10')\n",
    "barFPKMs('FPKM_table.tab', 'DDR2')\n",
    "barFPKMs('FPKM_table.tab', 'HSP12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 1.b##\n",
    "\n",
    "# Function to parse and plot the data\n",
    "def scatterFPKMs(tableFile,sample1,sample2):\n",
    "    # Figure out which columns we need.\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    i1 = header.index(sample1)\n",
    "    i2 = header.index(sample2)\n",
    "    \n",
    "    # Pull out the data series from each column.\n",
    "    # Keep genes that vary significantly in sigx and sigy.\n",
    "    # Keep less variable genes in insigx and insigy.\n",
    "    sigx = []\n",
    "    sigy = []\n",
    "    insigx = []\n",
    "    insigy = []\n",
    "    for line in fh:\n",
    "        line = line.strip().split()\n",
    "        x = float(line[i1])\n",
    "        y = float(line[i2])\n",
    "        if (x > 2 * y) or (y > 2 * x):\n",
    "            sigx.append(x)\n",
    "            sigy.append(y)\n",
    "        else:\n",
    "            insigx.append(x)\n",
    "            insigy.append(y)\n",
    "    \n",
    "    # Make a plot\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(1,1,1)\n",
    "    # Set log scales\n",
    "    ax.set_yscale('log')\n",
    "    ax.set_xscale('log')\n",
    "    # Set limits on the axes\n",
    "    # Because the scale is log, min must be >0\n",
    "    ax.set_xlim(1, 100000)\n",
    "    ax.set_ylim(1, 100000)\n",
    "    \n",
    "    # Plot the points from each line of data\n",
    "    # Significant variation in red\n",
    "    ax.plot(sigx, sigy, 'ro', alpha=0.2)\n",
    "    # Insignificant variation in black\n",
    "    ax.plot(insigx, insigy, 'ko', alpha=0.2)\n",
    "        \n",
    "    plt.show()\n",
    " \n",
    "scatterFPKMs(\"FPKM_table.tab\",'WT','sir2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 1.c##\n",
    "\n",
    "# Function to parse and plot the data\n",
    "# in this case, the samples plotted are supplied in a list\n",
    "def histogramFPKMs(tableFile, sample1, sample2):\n",
    "    # Figure out which columns we need.\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    idx1 = header.index(sample1)\n",
    "    idx2 = header.index(sample2)\n",
    "    \n",
    "    # Parse out the relevant columns and calculate the expression ratio\n",
    "    ratios = []\n",
    "    for line in fh:\n",
    "        line = line.strip().split()\n",
    "        fpkm1 = float(line[idx1])\n",
    "        fpkm2 = float(line[idx2])\n",
    "        if fpkm2: # cannot divide by zero\n",
    "            ratio = fpkm1 / fpkm2\n",
    "            ratios.append(ratio)\n",
    "    \n",
    "    # Make a plot & plot the data as a histogram\n",
    "    fig = plt.figure(figsize=(10,4))\n",
    "    ax = fig.add_subplot(1,1,1)\n",
    "    ax.hist(ratios, bins=50, range=(0,5), color='gray')\n",
    "        \n",
    "    plt.show()\n",
    " \n",
    "histogramFPKMs(\"FPKM_table.tab\", 'WT', 'sir3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 1.d##\n",
    "\n",
    "# Function to make beautiful bar graphs\n",
    "def barFPKMs(tableFile, gene):\n",
    "    # Grab the header\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    \n",
    "    # Grab the data for our gene\n",
    "    # The gene names are in the first column\n",
    "    for line in fh:\n",
    "        if line.startswith(gene):\n",
    "            data = line.strip().split()[1:]\n",
    "            data = map(float, data) # Convert all data to a float\n",
    "            break\n",
    "    else:\n",
    "        print \"Gene not found\"\n",
    "        return 0\n",
    "    \n",
    "    # Make a plot\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(1, 1, 1)\n",
    "    \n",
    "    # Matplotlib assumes you have numerical data for both axes\n",
    "    # But for the X axis, we have categorical data\n",
    "    # So we need to plug in a range of numbers to place the bars at\n",
    "    x = range(len(data))\n",
    "    \n",
    "    # Bars will appear at 0, 1, 2, etc\n",
    "    # So set the x-limits to include these values\n",
    "    ax.set_xlim(-1, len(data))\n",
    "    \n",
    "    # Here we turn the bars grey and remove their edge border.\n",
    "    # Also make the bars a little more narrow.\n",
    "    ax.bar(x, data, align='center', color='gray', edgecolor='none', width=0.5)\n",
    "    \n",
    "    # Now we change the labels for our categorical data\n",
    "    labels = [''] + header[1:] # We don't need to label 'Gene' at -1\n",
    "    ax.set_xticklabels(labels)\n",
    "    \n",
    "    # Hide the frame around the plot\n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    # Turn off the ticks\n",
    "    ax.tick_params(bottom='off', top='off', left='off', right='off')\n",
    "    # Overlay a white grid on the y axis\n",
    "    ax.yaxis.grid(True, color='white', linestyle='solid')\n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "barFPKMs('FPKM_table.tab', 'HMLALPHA2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 1.e##\n",
    "\n",
    "# Function to plot a beautiful fpkm bar graph on an axes object\n",
    "def barFPKMs(ax, tableFile, gene):\n",
    "    # Grab the header\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    \n",
    "    # Grab the data for our gene\n",
    "    # The gene names are in the first column\n",
    "    for line in fh:\n",
    "        if line.startswith(gene):\n",
    "            data = line.strip().split()[1:]\n",
    "            data = map(float, data) # Convert all data to a float\n",
    "            break\n",
    "    else:\n",
    "        print \"Gene not found\"\n",
    "        return 0\n",
    "    \n",
    "    # Matplotlib assumes you have numerical data for both axes\n",
    "    # But for the X axis, we have categorical data\n",
    "    # So we need to plug in a range of numbers to place the bars at\n",
    "    x = range(len(data))\n",
    "    \n",
    "    # Bars will appear at 0, 1, 2, etc\n",
    "    # So set the x-limits to include these values\n",
    "    ax.set_xlim(-1, len(data))\n",
    "    \n",
    "    # Here we turn the bars grey and remove their edge border.\n",
    "    # Also make the bars a little more narrow.\n",
    "    ax.bar(x, data, align='center', color='gray', edgecolor='none', width=0.5)\n",
    "    \n",
    "    # Now we change the labels for our categorical data\n",
    "    labels = [''] + header[1:] # We don't need to label 'Gene' at -1\n",
    "    ax.set_xticklabels(labels)\n",
    "    \n",
    "    # Hide the frame around the plot\n",
    "    for spine in ax.spines:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    # Turn off the ticks\n",
    "    ax.tick_params(bottom='off', top='off', left='off', right='off')\n",
    "    # Overlay a white grid on the y axis\n",
    "    ax.yaxis.grid(True, color='white', linestyle='solid')\n",
    "    \n",
    "    # Label the gene name on the subplot\n",
    "    ax.set_title(gene, fontweight='bold')\n",
    "\n",
    "# Make a figure with 4 axes subplots\n",
    "fig = plt.figure(figsize=(8,8))\n",
    "axes = [fig.add_subplot(2, 2, i) for i in range(1,5)]\n",
    "\n",
    "# Use the plotting function 4 times, ploting each gene onto each axes\n",
    "genes = ['HMLALPHA2', 'MAK10', 'DDR2', 'HSP12']\n",
    "[barFPKMs(axes[i], 'FPKM_table.tab', gene) for i, gene in enumerate(genes)]\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 2##\n",
    "\n",
    "#!/usr/env/bin python\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def parse_fastq(fastq_file):\n",
    "    fh = open(fastq_file, 'r')\n",
    "    \n",
    "    while True:\n",
    "        ID = fh.readline().strip()\n",
    "        seq = fh.readline().strip()\n",
    "        ID2 = fh.readline().strip()\n",
    "        qual = fh.readline().strip()\n",
    "        \n",
    "        if not ID:\n",
    "            break # end of file, finished successfully\n",
    "    \n",
    "        yield ID, seq, ID2, qual\n",
    "\n",
    "def base_frequencies(fastq_file):\n",
    "    fastq = parse_fastq(fastq_file)\n",
    "    \n",
    "    bases = ['A', 'T', 'C', 'G', 'N']\n",
    "    \n",
    "    # First initialize some arrays of the sequence length\n",
    "    for ID, seq, ID2, qual in fastq:    \n",
    "        seq = np.array(list(seq))\n",
    "        readlength = len(seq)\n",
    "        base_freqs = {base : (seq == base).astype(int) for base in bases}\n",
    "        break\n",
    "    \n",
    "    # fastq is an iterator, and like filehandles,\n",
    "    # iterators 'remember' their place\n",
    "    for ID, seq, ID2, qual in fastq:    \n",
    "        seq = np.array(list(seq))\n",
    "        for base in bases:\n",
    "            base_freqs[base] += seq == base\n",
    "            \n",
    "    # Find out the total number of reads\n",
    "    totals = sum(base_freqs.values()).astype(float)\n",
    "    \n",
    "    # Calculate the percent frequencies of each base at all positions\n",
    "    base_percents = {base : (base_freqs[base] / totals) for base in bases}\n",
    "    \n",
    "    return base_percents, readlength\n",
    "\n",
    "def base_linegraph(base_percents, readlength):\n",
    "    # Plot the percent of each base by read position\n",
    "    fig = plt.figure(figsize=(6,6))\n",
    "    ax = fig.add_subplot(1, 1, 1)\n",
    "    \n",
    "    # Note that base_percents is a\n",
    "    # dictionary keyed by the bases.\n",
    "    for base, percents in base_percents.items():\n",
    "        ax.plot(range(1, readlength+1), percents, label=base)\n",
    "\n",
    "    # Add titles as per b)\n",
    "    fig.suptitle('Read Biases')\n",
    "    ax.set_ylabel('% Nucleotide')\n",
    "    ax.set_xlabel('Read Position')\n",
    "\n",
    "    # Add a legend as per b)\n",
    "    legend = ax.legend(loc='lower right')\n",
    "    \n",
    "    plt.savefig('base_frequencies.png')\n",
    "\n",
    "    \n",
    "# Parse the fastq file\n",
    "base_percents, readlength = base_frequencies('illumina_reads.fastq')\n",
    "# Use our function to plot the data\n",
    "base_linegraph(base_percents, readlength)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 3##\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Function to parse and plot the data on an axes object\n",
    "def scatterFPKMs(axes, tableFile, sample1, sample2):\n",
    "    # Figure out which columns we need.\n",
    "    fh = open(tableFile, 'r')\n",
    "    header = fh.readline().strip().split()\n",
    "    i1 = header.index(sample1)\n",
    "    i2 = header.index(sample2)\n",
    "    \n",
    "    # Pull out the data series from each column\n",
    "    x = []\n",
    "    y = []\n",
    "    for line in fh:\n",
    "        line = line.strip().split()\n",
    "        x.append(line[i1])\n",
    "        y.append(line[i2])\n",
    "    \n",
    "    fh.close()\n",
    "    \n",
    "    # Plot all the points\n",
    "    axes.plot(x, y, 'bo')\n",
    "\n",
    "    # Set log scale and limits\n",
    "    axes.set_yscale('log')\n",
    "    axes.set_xscale('log')\n",
    "    axes.set_xlim(1, 100000)\n",
    "    axes.set_ylim(1, 100000)\n",
    "    \n",
    "    # Add labels\n",
    "    axes.set_xlabel(sample1)\n",
    "    axes.set_ylabel(sample2)\n",
    "    \n",
    "table = 'FPKM_table.tab'\n",
    "\n",
    "fig = plt.figure(figsize=(9,3))\n",
    "axeses = [fig.add_subplot(1, 3, i+1) for i in range(3)]\n",
    "\n",
    "scatterFPKMs(axeses[0], table, 'WT', 'hoD')\n",
    "scatterFPKMs(axeses[1], table, 'WT', 'sir4DK')\n",
    "scatterFPKMs(axeses[2], table, 'hoD', 'sir4DK')\n",
    "\n",
    "fig.tight_layout() # This will make the labels fit better\n",
    "plt.savefig('exercise3.ps')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##Exercise 4##\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "\n",
    "def barplot_with_error(tabfile):\n",
    "    # Parse the table\n",
    "    table = open(tabfile, 'r')\n",
    "\n",
    "    header = table.readline()\n",
    "\n",
    "    conditions = []\n",
    "    means = []\n",
    "    sems = []\n",
    "    for line in table:\n",
    "        line = line.strip().split()\n",
    "        condition = line[0]\n",
    "        reps = map(float, line[1:])\n",
    "        mean = np.mean(reps)\n",
    "        sem = stats.sem(reps, ddof=len(reps)-1)\n",
    "    \n",
    "        conditions.append(condition)\n",
    "        means.append(mean)\n",
    "        sems.append(sem)\n",
    "    \n",
    "    # Plot the data in a bar graph\n",
    "    fig = plt.figure(figsize=(6,6))\n",
    "    ax = fig.add_subplot(1,1,1)\n",
    "\n",
    "    x = range(len(conditions))\n",
    "    ax.set_xlim(-1, len(x))\n",
    "    ax.bar(x, means, align='center', yerr=sems)\n",
    "    ax.set_xticklabels([''] + conditions)\n",
    "\n",
    "    plt.show()\n",
    "    \n",
    "barplot_with_error('exercise4.tab')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
